### PR TITLE
Add validation for deal POST, endpoint for updating deal, and tests for deal POST servlet

### PR DESCRIPTION
**Pull request by limli**

### CODE REVIEW

File: src/main/java/com/google/step/servlets/DealDetailServlet.java [hunk 1]
```diff
@@ -52,4 +54,56 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
     response.getWriter().println(JsonFormatter.getDealJson(deal));
   }
+
+  /** Updates the deal with the given id parameter */
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    long id;
+    try {
+      id = Long.parseLong(request.getPathInfo().substring(1));
+    } catch (NumberFormatException e) {
```
Comment (jyx11011) [inline src/main/java/com/google/step/servlets/DealDetailServlet.java]: can also catch an index out of bound exception thrown by substring method here

**Comment by SuyashLakhotia** [comment]
Please add tests! :)

**Review by SuyashLakhotia — CHANGES_REQUESTED** [review]
Please add the appropriate tests for the servlet.

**Comment by limli** [comment]
> Please add the appropriate tests for the servlet.

@SuyashLakhotia the tests are inside another branch https://github.com/googleinterns/step203-2020/tree/deal-test for json test and https://github.com/googleinterns/step203-2020/tree/deal-test-servlet for servlet test, which is checkout-ed from branch user-test. Should I merge the branches to here?

**Comment by SuyashLakhotia** [comment]
@limli Thanks for the pointers! I think we should coordinate with @jyx11011 to get her PRs that you're dependent on merged into `master` so that you can add the tests in this PR alongside the code and merge it together.

However, don't merge `user-test` into this branch (unless it's already merged with `master`) because that will make the diff super hard to read.

**Comment by limli** [comment]
@SuyashLakhotia the POST servlet tests are added. The tests for the other servlets, as well as the tests for the JsonFormatter and datastore can be done in another PR.

File: src/main/java/com/google/step/servlets/DealDetailServlet.java [hunk 2]
```diff
@@ -52,4 +54,56 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    } catch (NumberFormatException | IndexOutOfBoundsException e) {
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+    String description = request.getParameter("description");
+    String photoBlobkey = null; // TODO connect to blobstore
+    String start = request.getParameter("start");
+    String end = request.getParameter("end");
+    String source = request.getParameter("source");
+    long poster = -1;
```
Comment (SuyashLakhotia) [inline src/main/java/com/google/step/servlets/DealDetailServlet.java]:
```suggestion
    long posterId = -1;
```

File: src/main/java/com/google/step/servlets/DealDetailServlet.java [hunk 3]
```diff
@@ -52,4 +54,56 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    long restaurant = -1;
```
Comment (SuyashLakhotia) [inline src/main/java/com/google/step/servlets/DealDetailServlet.java]:
```suggestion
    long restaurantId = -1;
```

File: src/main/java/com/google/step/servlets/DealPostServlet.java [hunk 4]
```diff
@@ -15,27 +17,71 @@
...
   /** Posts the deal with the given id parameter */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    System.out.println("post deal");
     String description = request.getParameter("description");
     String photoBlobkey = "TODO"; // TODO connect to blobstore
     String start = request.getParameter("start");
     String end = request.getParameter("end");
     String source = request.getParameter("source");
     long poster = 1234; // TODO get authenticated user
```
Comment (SuyashLakhotia) [inline src/main/java/com/google/step/servlets/DealPostServlet.java]:
```suggestion
    long posterId = 1234; // TODO get authenticated user
```

File: src/main/java/com/google/step/servlets/DealPostServlet.java [hunk 5]
```diff
@@ -15,27 +17,71 @@
...
-    long restaurant = Long.parseLong(request.getParameter("restaurant"));
-    // TODO validate all entries
+    long restaurant;
```
Comment (SuyashLakhotia) [inline src/main/java/com/google/step/servlets/DealPostServlet.java]:
```suggestion
    long restaurantId;
```

File: src/test/java/com/google/step/servlets/DealPostServletTest.java [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+  private DealManager dealManager;
+
+  @Before
+  public void setUp() {
+    dealManager = mock(DealManager.class);
+    servlet = new DealPostServlet(dealManager);
+  }
+
+  @Test
+  public void testDoPost_sucess() throws IOException {
```
Comment (SuyashLakhotia) [inline src/test/java/com/google/step/servlets/DealPostServletTest.java]:
```suggestion
  public void testDoPost_success() throws IOException {
```

File: src/test/java/com/google/step/servlets/DealPostServletTest.java [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+            eq(DATE_A),
+            eq(DATE_B),
+            eq(SOURCE_A),
+            anyLong(),
+            eq(RESTAURANT_ID_A)))
+        .thenReturn(DEAL);
+
+    servlet.doPost(request, response);
+
+    verify(response, never()).setStatus(HttpServletResponse.SC_BAD_REQUEST);
```
Comment (SuyashLakhotia) [inline src/test/java/com/google/step/servlets/DealPostServletTest.java]: Verify status is `SC_OK`? Add this to the servlet if not already the case.

Comment (SuyashLakhotia) [reply src/test/java/com/google/step/servlets/DealPostServletTest.java]: You should also verify that the DealManager is being invoked correctly here.

**Review by SuyashLakhotia — CHANGES_REQUESTED** [review]

**Review by SuyashLakhotia — DISMISSED** [review]

File: src/test/java/com/google/step/servlets/DealPostServletTest.java [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+    verify(dealManager)
+        .createDeal(
+            eq(DESCRIPTION_A),
+            anyString(),
+            eq(DATE_A),
+            eq(DATE_B),
+            eq(SOURCE_A),
+            anyLong(),
+            eq(RESTAURANT_ID_A));
+    verify(response, never()).setStatus(HttpServletResponse.SC_BAD_REQUEST);
```
Comment (SuyashLakhotia) [inline src/test/java/com/google/step/servlets/DealPostServletTest.java]: You can remove this line now.

**Review by SuyashLakhotia — APPROVED** [review]

**Pull request merged**